### PR TITLE
[Choreo - WebsSocket Feature] Update wasm module to fix memory leaks

### PR DESCRIPTION
**Pull request by Thushani-Jayasekera**
### Purpose
To update files in `envoy-filters/mgw-source/filters/http/mgw-wasm-websocket` of the Choreo branch, similar to the main branch.

PR with fixes to main branch: https://github.com/wso2/product-microgateway/commit/61ada7ec58f4040d7f2eb64456dee400579c6a32
### Issues
Fixes #

### Automation tests
 - Unit tests added: Yes/No
 - Integration tests added: Yes/No

### Tested environments

Not Tested

---
#### Maintainers: Check before merge
- [x] Assigned 'Type' label
- [ ] Assigned the project
- [x] Validated respective github issues
- [x] Assigned milestone to the github issue(s)

### CODE REVIEW

File: envoy-filters/mgw-source/filters/http/mgw-wasm-websocket/BUILD [hunk 1]
```diff
@@ -4,8 +4,7 @@ wasm_cc_binary(
     name = "mgw-websocket.wasm",
     srcs = ["filter.cc", "filter.h", "handler_impl.h", "handler_impl.cc", "handler.h"],
     deps = [
-        "@proxy_wasm_cpp_sdk//:proxy_wasm_intrinsics",
         "@mgw_wasm_websocket_api//:api_cc_proto",
-        "@proxy_wasm_cpp_sdk//:proxy_wasm_intrinsics_lite"
+        "@proxy_wasm_cpp_sdk//:proxy_wasm_intrinsics_full"
```
Comment (Thushani-Jayasekera) [inline envoy-filters/mgw-source/filters/http/mgw-wasm-websocket/BUILD:8]: @VirajSalaka pls confirm on this dependency update

File: envoy-filters/mgw-source/filters/http/mgw-wasm-websocket/filter.cc [hunk 2]
```diff
@@ -86,24 +94,71 @@ void MgwWebSocketContext::onCreate() {
...
+      getValue({"upstream", "address"}, &upstream_address);
+      WebSocketFrameRequest request;
+      request.set_node_id(this->node_id_);
+      request.set_frame_length(0);
+      request.set_remote_ip(upstream_address);
+      // Read ext_authz_metadata_ metdata saved as a member variable
+      *request.mutable_metadata() = *this->metadata_;
+      request.set_payload("");
+      request.set_direction(WebSocketFrameRequest_MessageDirection_HANDSHAKE);
+      request.set_apim_error_code(0);
```
Comment (Thushani-Jayasekera) [inline envoy-filters/mgw-source/filters/http/mgw-wasm-websocket/filter.cc:157]: Pls check on setting apim error code

File: envoy-filters/mgw-source/filters/http/mgw-wasm-websocket/filter.cc [hunk 3]
```diff
@@ -248,14 +276,21 @@ FilterDataStatus MgwWebSocketContext::onResponseBody(size_t body_buffer_length,
...
           return FilterDataStatus::Continue;
         }else{
+          sendEnforcerRequest(this, request);
           return FilterDataStatus::StopIterationNoBuffer;
         }
       }else{
+        // It is unlikely that the return value would be zero https://man7.org/linux/man-pages/man2/gettimeofday.2.html
+        LOG_ERROR(std::string("Current Time cannot be processed. Hence the websocket stream is closed. : ") +
+            this->x_request_id_);
+        return FilterDataStatus::StopIterationNoBuffer;
```
Comment (VirajSalaka) [inline envoy-filters/mgw-source/filters/http/mgw-wasm-websocket/filter.cc:293]: same line repeated twice.

File: envoy-filters/mgw-source/filters/http/mgw-wasm-websocket/filter.cc [hunk 4]
```diff
@@ -86,24 +91,70 @@ void MgwWebSocketContext::onCreate() {
...
+    if (std::string(p.first) == X_REQUEST_ID) {
+      this->x_request_id_ = std::string(p.second);
+    }
+  }
   // Extract ext_authz dynamic metadata and assign it to a member variable 
   auto buffer = getProperty<std::string>({"metadata", "filter_metadata", "envoy.filters.http.ext_authz"});
   if (buffer.has_value() && buffer.value()->size() != 0) {
     auto pairs = buffer.value()->pairs();
     for (auto &p : pairs) {
+      // TODO (thushani) Implement fixes for throttling
```
Comment (VirajSalaka) [inline envoy-filters/mgw-source/filters/http/mgw-wasm-websocket/filter.cc:106]: shall we omit this logic from here? It feels like we need an entirely new way to handle throttling as now enforcer does not perform throttling .

File: envoy-filters/mgw-source/filters/http/mgw-wasm-websocket/filter.cc [hunk 5]
```diff
@@ -113,14 +164,13 @@ FilterHeadersStatus MgwWebSocketContext::onResponseHeaders(uint32_t, bool) {
 FilterDataStatus MgwWebSocketContext::onRequestBody(size_t body_buffer_length,
                                                bool /* end_of_stream */) {
   auto body = getBufferBytes(WasmBufferType::HttpRequestBody, 0, body_buffer_length);
-  LOG_TRACE(std::string("onRequestBody called mgw_WASM_websocket") + std::string(body->view()));
+  LOG_TRACE(std::string("onRequestBody called mgw_WASM_websocket ") + std::string(body->view()) + std::string(" : ") + this->x_request_id_ );
```
Comment (VirajSalaka) [inline envoy-filters/mgw-source/filters/http/mgw-wasm-websocket/filter.cc:167]: shall we not include the body->view all together? I don't think we need to print the payload.

File: envoy-filters/mgw-source/filters/http/mgw-wasm-websocket/filter.cc [hunk 6]
```diff
@@ -335,5 +372,19 @@ void MgwWebSocketContext::establishNewStream() {
...
+
+void MgwWebSocketContext::sendEnforcerRequest(MgwWebSocketContext* websocContext, WebSocketFrameRequest request) {
+  if (websocContext->handler_state_ == HandlerState::OK) {
+    LOG_TRACE(std::string("gRPC bidi stream available. publishing frame data...") + std::string(" : ") + this->x_request_id_);
+  } else {
+    establishNewStream();
+  }
+  auto ack = websocContext->stream_handler_->send(request, false);
+  if (ack != WasmResult::Ok) {
+    LOG_WARN(std::string("error sending frame data")+ toString(ack) + std::string(" : ") + this->x_request_id_);
```
Comment (VirajSalaka) [inline envoy-filters/mgw-source/filters/http/mgw-wasm-websocket/filter.cc:386]: when you are testing let's look for this log too.

File: envoy-filters/mgw-source/filters/http/mgw-wasm-websocket/handler_impl.cc [hunk 7]
```diff
@@ -50,11 +51,11 @@ void MgwGrpcStreamHandler::onRemoteClose(GrpcStatus status){
   this->callbacks_->updateHandlerState(HandlerState::Error);
 };
 
-bool MgwGrpcStreamHandler::sendMessage(WebSocketFrameRequest request){
+void MgwGrpcStreamHandler::sendMessage(WebSocketFrameRequest request){
   auto res = send(request, true);
   if(res != WasmResult::Ok){
-    return false;
+    LOG_TRACE(std::string("Error sending message"));
```
Comment (VirajSalaka) [inline envoy-filters/mgw-source/filters/http/mgw-wasm-websocket/handler_impl.cc:57]: This looks like a warning log.

File: envoy-filters/mgw-source/filters/http/mgw-wasm-websocket/BUILD (hunk 1 above)
Comment (VirajSalaka) [reply envoy-filters/mgw-source/filters/http/mgw-wasm-websocket/BUILD:8]: Can't exactly recall why it was added previously. As long as it builds and does the job, we are good. This does not affect regular REST calls.

File: envoy-filters/mgw-source/filters/http/mgw-wasm-websocket/filter.cc (hunk 6 above)
Comment (Thushani-Jayasekera) [reply envoy-filters/mgw-source/filters/http/mgw-wasm-websocket/filter.cc:386]: Ack

File: envoy-filters/mgw-source/filters/http/mgw-wasm-websocket/filter.cc [hunk 8]
```diff
@@ -53,21 +53,26 @@ bool MgwWebSocketRootContext::onConfigure(size_t config_size) {
...
+  if (!options_status.ok()) {
     LOG_WARN("Cannot parse plugin configuration JSON string: " + configuration->toString());
     return false;
   }
   LOG_TRACE("Loading Config: " + config_.node_id());
   return true;
 }
 
+MgwWebSocketContext::~MgwWebSocketContext() {
+  LOG_TRACE(std::string("MgwContext destructed") + this->x_request_id_);
```
Comment (renuka-fernando) [inline envoy-filters/mgw-source/filters/http/mgw-wasm-websocket/filter.cc:65]:
```suggestion
  LOG_TRACE(std::string("MgwContext destructed: ") + this->x_request_id_);
```

File: envoy-filters/mgw-source/filters/http/mgw-wasm-websocket/filter.cc [hunk 9]
```diff
@@ -113,14 +143,13 @@ FilterHeadersStatus MgwWebSocketContext::onResponseHeaders(uint32_t, bool) {
 FilterDataStatus MgwWebSocketContext::onRequestBody(size_t body_buffer_length,
                                                bool /* end_of_stream */) {
   auto body = getBufferBytes(WasmBufferType::HttpRequestBody, 0, body_buffer_length);
-  LOG_TRACE(std::string("onRequestBody called mgw_WASM_websocket") + std::string(body->view()));
+  LOG_TRACE(std::string("onRequestBody called mgw_WASM_websocket ") + std::string(" : ") + this->x_request_id_ );
```
Comment (renuka-fernando) [inline envoy-filters/mgw-source/filters/http/mgw-wasm-websocket/filter.cc:146]:
```suggestion
  LOG_TRACE(std::string("onRequestBody called mgw_WASM_websocket : ") + this->x_request_id_ );
```

**Review by renuka-fernando — APPROVED** [review]
Let's merge this as this fix won't affect non WS API calls.
If @VirajSalaka has further comments we can handle those in a new PR.

**Comment by choreo-cicd** [comment]
[succeeded] Dataplane(NorthEU) cluster : dev-deployment-v2 : 20240603.11

**Comment by choreo-cicd** [comment]
[succeeded]  : dev-deployment-v2 : 20240603.11

**Comment by choreo-cicd** [comment]
[succeeded] Controlplane cluster : dev-deployment-v2 : 20240603.11

**Comment by choreo-cicd** [comment]
[succeeded] Dataplane(EastUS) cluster : dev-deployment-v2 : 20240603.11

**Comment by choreo-cicd** [comment]
[] Dataplane(EastUS) cluster : stage-deployment-v2 : 20240606.2

**Comment by choreo-cicd** [comment]
[] Dataplane(NorthEU) cluster : stage-deployment-v2 : 20240606.2

**Comment by choreo-cicd** [comment]
[] Dataplane(CDP2-WU2) cluster : stage-deployment-v2 : 20240606.2

**Comment by choreo-cicd** [comment]
[] Controlplane cluster : stage-deployment-v2 : 20240606.2

**Comment by choreo-cicd** [comment]
[succeeded] Dataplane(CDP2-WU2) cluster : prod-deployment-v2 : 20240607.2

**Comment by choreo-cicd** [comment]
[failed] Controlplane cluster : prod-deployment-v2 : 20240607.2

**Comment by choreo-cicd** [comment]
[succeeded] Dataplane(NorthEU) cluster : prod-deployment-v2 : 20240607.2

**Comment by choreo-cicd** [comment]
[succeeded] Dataplane(CDP2-WU2) cluster : prod-deployment-v2 : 20240607.2

**Comment by choreo-cicd** [comment]
[succeeded] Controlplane cluster : prod-deployment-v2 : 20240607.2

**Comment by choreo-cicd** [comment]
[succeeded] Dataplane(EastUS) cluster : prod-deployment-v2 : 20240607.2

**Pull request merged**